File: Sources/Shells.swift
import Foundation

/// The commands a session left running in the background, which its screen stops saying out loud.
///
/// This is the other half of ``Subagents``, and it exists for the same reason: work moved
/// somewhere Claude Code does not draw. A background shell is worse than an agent on that count,
/// because the terminal *does* mention it — once. "Cooked for 1h 25m 13s · 1 shell still running"
/// is printed where the turn ended, above an ordinary prompt, and it scrolls away the moment
/// anything else is said. Everywhere this app looks after that says the same thing an ordinary
/// finished session says: nothing running, nothing asked, idle. **A session with a build still
/// going reads as a session that is done**, which is the one wrong answer a fleet list must not
/// give, and it is the whole reason this file exists.
///
/// The facts are on disk. Every `Bash` call gets an output file of its own, under a directory
/// named for the session:
///
/// ```
/// /tmp/claude-<uid>/<project>/<session>/tasks/
///     <task-id>.output   ← appended to while the command runs
/// ```
///
/// **Two shapes tell the whole story, and both were observed rather than documented.** A command
/// that has finished in the background has `[exited with code 0]` written under its last line and
/// keeps its file; a command running in the foreground has its file *deleted* the moment it
/// returns. So a file with no exit marker under it is a command that has not finished — and the
/// only way it can be a foreground one is if the session is working, which is the state this
/// reading refuses to answer for. See ``reading(of:states:)``.
///
/// Nothing here is promised by anybody and all of it can change. Same rule as everywhere else:
/// recognise a shape, and a session whose files say nothing recognisable has no shells rather
/// than an error.
enum Shells {

    /// One command still running where nobody can see it.
    struct Shell: Equatable {
        /// Claude Code's own id for it — `bf7e0p5v7` — which is also the name of its file and
        /// the word `/bashes` uses for it on the terminal.
        let id: String
        /// When it last printed something. For a command that prints as it goes this is a live
        /// clock; for `sleep 600` it is when it started, and that is the honest answer.
        let at: Date
        /// The last line it printed. The closest thing to a live line something with no screen
        /// has, and the same slot ``Subagents/Agent/doing`` fills for an agent.
        let doing: String?
    }

    /// How long after its last write a command is still worth reporting.
    ///
    /// Not a guess at how long a build takes — it is how long a *silent* one stays believable. A
    /// command that prints nothing for twelve hours is either finished in a way that left no
    /// marker (Claude Code killed mid-run, the machine slept through it) or it is the rarest
    /// thing on this list, and a chip that never goes away is worse than a chip that is late.
    private static let liveWindow: TimeInterval = 12 * 60 * 60

    /// At most this many per session, newest first. `/bashes` will list thirty; a strip above a
    /// transcript can show a few, and the honest thing to do with the rest is say how many.
    static let shown = 6

    /// The size of the tail read from each output file. Big enough for the last line of anything
    /// that prints normally, small enough that a session with six of them is not doing I/O
    /// anybody would notice once a second.
    private static let tailBytes = 8 << 10

    /// Every session's running background commands, keyed by session id. Sessions with none are
    /// absent.
    ///
    /// **A working session is skipped, and that is a correctness rule rather than a saving.** The
    /// file a foreground `Bash` is writing to right now looks exactly like the file a background
    /// one is writing to right now — no marker under either, because neither has finished. What
    /// separates them is the session: a foreground command *is* the session being busy, so
    /// nothing that is running in the foreground can belong to a session that is not working.
    /// Asking only the sessions that are between turns removes the one ambiguity these files
    /// have, and it removes it in the only state this reading is for. A session that is working
    /// already looks like a session that is working.
    ///
    /// Claude Code only, and `isClaude` rather than `isAssistant` says so on purpose: Codex keeps
    /// no such directory, so asking would be a `stat` that can only come back empty, once per
    /// session per beat.
    static func reading(of sessions: [TargetSession],
                        states: [String: SessionState]) -> [String: [Shell]] {
        var out: [String: [Shell]] = [:]
        for session in sessions where session.isClaude {
            if case .working = states[session.id] ?? .unknown { continue }
            let found = shells(of: session)
            if !found.isEmpty { out[session.id] = found }
        }
        return out
    }

    /// **The fast path is the one with nothing running, and it is nearly all of them.**
    ///
    /// A session between turns has had its foreground output files deleted behind it and has no
    /// background ones, so its `tasks` directory is empty and this costs one `readdir`. Only a
    /// session that actually has a file in there ever reads a byte.
    static func shells(of session: TargetSession) -> [Shell] {
        guard let folder = folder(of: session) else { return [] }
        let fm = FileManager.default
        guard let names = try? fm.contentsOfDirectory(atPath: folder.path) else { return [] }

        let now = Date()
        var shells: [Shell] = []
        for name in names where name.hasSuffix(".output") {
            let id = String(name.dropLast(".output".count))
            guard !id.isEmpty else { continue }
            let file = folder.appendingPathComponent(name)
            guard let at = modified(file), now.timeIntervalSince(at) < liveWindow else { continue }
            let tail = read(file, signature: "\(at.timeIntervalSince1970)-\(size(file))")
            guard !tail.ended else { continue }
            shells.append(Shell(id: id, at: at, doing: tail.last))
        }

        // Newest first: the one that printed a second ago is the one somebody is waiting on.
        shells.sort { $0.at > $1.at }
        return Array(shells.prefix(shown))
    }

    // MARK: - Where the files are

    /// Where Claude Code keeps a session's command output, whether or not anything is in there.
    ///
    /// Derived from the transcript rather than guessed at, because the two paths are built from
    /// the same two names: `~/.claude/projects/<project>/<session>.jsonl` beside
    /// `/tmp/claude-<uid>/<project>/<session>/tasks`. Finding the transcript is
    /// ``Subagents/transcript(of:)``'s job and it is already cached per session, so this adds no
    /// directory listing of its own.
    ///
    /// `/tmp` literally, not `NSTemporaryDirectory()`: on macOS that answers the per-process
    /// `/var/folders/…` sandbox, which is not where the writer of these files put them.
    static func folder(of session: TargetSession) -> URL? {
        guard let transcript = Subagents.transcript(of: session) else { return nil }
        let name = transcript.deletingPathExtension().lastPathComponent
        let project = transcript.deletingLastPathComponent().lastPathComponent
        guard !name.isEmpty, !project.isEmpty else { return nil }
        return URL(fileURLWithPath: "/tmp/claude-\(getuid())", isDirectory: true)
            .appendingPathComponent(project, isDirectory: true)
            .appendingPathComponent(name, isDirectory: true)
            .appendingPathComponent("tasks", isDirectory: true)
    }

    // MARK: - What one file says

    /// What the tail of an output file says: whether the command has ended, and the last thing it
    /// printed before it did or did not.
    struct Ending: Equatable {
        var ended: Bool
        var last: String?
    }

    /// The marker Claude Code writes under a background command once it is over. Observed, and
    /// the one string in here that decides anything — so it is matched at the end of the file and
    /// nowhere else. A build that happens to print a line like this in the middle of its own
    /// output says nothing about whether it has finished, and this does not ask.
    private static let exitMarker = "[exited"

    /// Read once per version of the file. A command that has printed nothing since the last beat
    /// has the same size and the same mtime, and re-reading it would be eight kilobytes of I/O
    /// per session per second to arrive at the answer already held.
    private static let lock = NSLock()
    private static var reads: [String: (signature: String, ending: Ending)] = [:]

    static func read(_ url: URL, signature: String) -> Ending {
        lock.lock()
        if let hit = reads[url.path], hit.signature == signature {
            defer { lock.unlock() }
            return hit.ending
        }
        lock.unlock()

        var ending = Ending(ended: false, last: nil)
        if let text = Transcript.tail(of: url, bytes: tailBytes) {
            let lines = text
                .split(separator: "\n", omittingEmptySubsequences: false)
                .map { $0.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
            if let last = lines.last {
                // The marker is the last line or it is not the marker. Anything under it would be
                // output printed after the command ended, which cannot happen.
                if last.hasPrefix(exitMarker), last.hasSuffix("]") {
                    ending.ended = true
                } else {
                    ending.last = last
                }
            }
        }

        lock.lock()
        reads[url.path] = (signature, ending)
        lock.unlock()
        return ending
    }

    // MARK: - Odds and ends

    private static func modified(_ url: URL) -> Date? {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
        return attrs?[.modificationDate] as? Date
    }

    private static func size(_ url: URL) -> Int {
        let attrs = try? FileManager.default.attributesOfItem(atPath: url.path)
        return (attrs?[.size] as? Int) ?? 0
    }

    /// Everything remembered, let go of. Called alongside the other caches that only exist to
    /// make an open panel cheap.
    static func forget() {
        lock.lock()
        reads = [:]
        lock.unlock()
    }
}

File: internal/adapters/transcript/shells.go
package transcript

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
	"unicode"
	"unicode/utf8"

	"github.com/sainteye/clawdline/internal/domain/session"
)

// Shells finds the commands a Claude session left running in the background —
// the Swift app's `Shells`, reproduced rule for rule.
//
// Every `Bash` call gets an output file of its own:
//
//	/tmp/claude-<uid>/<project>/<session>/tasks/<task-id>.output
//
// **The file alone does not settle it.** A background command that finished
// has `[exited with code 0]` (or `[killed]`) written under its last line, and a
// foreground one has its file deleted when it returns — but a foreground one
// that was interrupted leaves its file behind with no marker, looking exactly
// like a build still going. So the transcript is the second fact: Claude Code
// answers a backgrounded call with "Command running in background with ID: …",
// and only an id announced that way, whose file has no ending under it, is a
// running command.
//
// Nothing here is promised by anybody. A session whose files say nothing this
// recognises has no shells, never an error.
type Shells struct {
	// Root is where Claude Code writes the output files: `/tmp` literally, not
	// the per-process temporary directory, which is not where the writer put
	// them.
	Root string
	UID  int
	Now  func() time.Time

	mu     sync.Mutex
	reads  map[string]shellRead
	starts map[string]announcements
}

type shellRead struct {
	signature string
	ended     bool
	last      string
}

type announcements struct {
	offset int64
	found  map[string]asked
}

type asked struct {
	command string
	what    string
}

const (
	// shellLiveWindow is how long a silent command stays believable. Past it,
	// it either ended without a marker or is the rarest thing on the list, and
	// a line that never goes away is worse than one that is late.
	shellLiveWindow = 12 * time.Hour
	// shellsShown is how many are reported per session, newest first.
	shellsShown = 6
	// shellTail is how much of an output file's end is read for its last line.
	shellTail = 8 << 10
	// shellRoom is how much of one line is carried: the width of a phone.
	shellRoom = 160

	backgroundAnnouncement = "Command running in background with ID: "
	backgroundFlag         = "run_in_background"
)

func NewShells() *Shells {
	return &Shells{
		Root:   "/tmp",
		UID:    os.Getuid(),
		Now:    time.Now,
		reads:  map[string]shellRead{},
		starts: map[string]announcements{},
	}
}

// Folder is where the session whose transcript is at transcriptPath keeps its
// command output. Built from the same two names the transcript path is.
func (s *Shells) Folder(transcriptPath string) string {
	name := strings.TrimSuffix(filepath.Base(transcriptPath), filepath.Ext(transcriptPath))
	project := filepath.Base(filepath.Dir(transcriptPath))
	if name == "" || project == "" || project == "." || project == string(filepath.Separator) {
		return ""
	}
	return filepath.Join(s.Root, fmt.Sprintf("claude-%d", s.UID), project, name, "tasks")
}

// Running is every background command still going for this transcript's
// session, newest first.
//
// The ordinary case is a session with nothing running, and it costs one
// directory listing: only a session with an unfinished file opens its
// transcript.
func (s *Shells) Running(transcriptPath string) []session.Shell {
	folder := s.Folder(transcriptPath)
	if folder == "" {
		return nil
	}
	entries, err := os.ReadDir(folder)
	if err != nil {
		return nil
	}
	now := s.Now()
	var unfinished []session.Shell
	for _, e := range entries {
		name := e.Name()
		if !strings.HasSuffix(name, ".output") {
			continue
		}
		id := strings.TrimSuffix(name, ".output")
		if id == "" {
			continue
		}
		info, err := e.Info()
		if err != nil {
			continue
		}
		at := info.ModTime()
		if now.Sub(at) >= shellLiveWindow {
			continue
		}
		file := filepath.Join(folder, name)
		read := s.read(file, fmt.Sprintf("%d-%d", at.UnixNano(), info.Size()))
		if read.ended {
			continue
		}
		unfinished = append(unfinished, session.Shell{ID: id, At: at, Doing: read.last})
	}
	if len(unfinished) == 0 {
		return nil
	}
	if _, err := os.Stat(transcriptPath); err != nil {
		return nil
	}
	announced := s.announced(transcriptPath)

	out := make([]session.Shell, 0, len(unfinished))
	for _, sh := range unfinished {
		was, ok := announced[sh.ID]
		if !ok {
			continue
		}
		sh.Command, sh.What = was.command, was.what
		out = append(out, sh)
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].At.After(out[j].At) })
	if len(out) > shellsShown {
		out = out[:shellsShown]
	}
	return out
}

// read is what one output file's tail says, once per version of the file.
func (s *Shells) read(path, signature string) shellRead {
	s.mu.Lock()
	if hit, ok := s.reads[path]; ok && hit.signature == signature {
		s.mu.Unlock()
		return hit
	}
	s.mu.Unlock()

	out := shellRead{signature: signature}
	if data, _, err := tailData(path, shellTail); err == nil {
		last := ""
		for _, line := range strings.Split(string(data), "\n") {
			if t := trimSpaces(line); t != "" {
				last = t
			}
		}
		switch {
		case last == "":
		case isEnding(last):
			out.ended = true
		default:
			out.last = clipped(last)
		}
	}

	s.mu.Lock()
	s.reads[path] = out
	s.mu.Unlock()
	return out
}

// isEnding is one of the two markers Claude Code writes under a background
// command once it is over — matched by its words, not by its brackets, because
// a program's own log line can be bracketed too.
func isEnding(line string) bool {
	return (strings.HasPrefix(line, "[exited") && strings.HasSuffix(line, "]")) || line == "[killed]"
}

// announced is every background command the transcript has announced, by id.
//
// Read forward from where the last read stopped, not from the end: the command
// still running after ninety minutes is the one whose announcement is furthest
// back. A file shorter than the offset was replaced, and is read again.
//
// Two records are joined by the tool call between them: the assistant's
// `tool_use`, marked `run_in_background`, carries the command, and the
// `tool_result` answering it carries the id Claude Code minted. Every line is
// tested for its substring before anything is decoded.
func (s *Shells) announced(path string) map[string]asked {
	s.mu.Lock()
	seen, had := s.starts[path]
	s.mu.Unlock()

	found := map[string]asked{}
	offset := int64(0)
	if had {
		offset = seen.offset
		for k, v := range seen.found {
			found[k] = v
		}
	}

	f, err := os.Open(path)
	if err != nil {
		return found
	}
	defer f.Close()
	size, err := f.Seek(0, io.SeekEnd)
	if err != nil {
		return found
	}
	if size < offset {
		offset = 0
		found = map[string]asked{}
	}
	if size <= offset {
		return found
	}
	if _, err := f.Seek(offset, io.SeekStart); err != nil {
		return found
	}

	// Only within this pass: a command whose two records straddle two reads
	// keeps its id and loses its command line, which is the right way round
	// to fail — the id is what decides whether anything is running.
	calls := map[string]asked{}
	r := bufio.NewReaderSize(io.LimitReader(f, size-offset), 256<<10)
	advanced := offset
	for {
		line, err := r.ReadBytes('\n')
		if err != nil {
			// A line with no newline yet is still being written; it is read
			// next time, whole.
			break
		}
		advanced += int64(len(line))
		if bytes.Contains(line, []byte(backgroundFlag)) {
			if rec, ok := decodeObject(line); ok {
				for _, block := range contentBlocks(rec) {
					if t, _ := block.str("type"); t != "tool_use" {
						continue
					}
					id, ok := block.str("id")
					if !ok {
						continue
					}
					input, ok := block.object("input")
					if !ok {
						continue
					}
					if bg, ok := input.boolean(backgroundFlag); !ok || !bg {
						continue
					}
					command, ok := input.str("command")
					if !ok {
						continue
					}
					was := asked{command: clipped(oneLine(command))}
					if what, ok := input.str("description"); ok {
						was.what = clipped(oneLine(what))
					}
					calls[id] = was
				}
			}
		}
		if !bytes.Contains(line, []byte(backgroundAnnouncement)) {
			continue
		}
		rec, ok := decodeObject(line)
		if !ok {
			continue
		}
		for _, block := range contentBlocks(rec) {
			if t, _ := block.str("type"); t != "tool_result" {
				continue
			}
			text, _ := block.str("content")
			i := strings.Index(text, backgroundAnnouncement)
			if i < 0 {
				continue
			}
			rest := text[i+len(backgroundAnnouncement):]
			n := 0
			for n < len(rest) && isASCIIAlnum(rest[n]) {
				n++
			}
			if n == 0 {
				continue
			}
			call, _ := block.str("tool_use_id")
			found[rest[:n]] = calls[call]
		}
	}

	s.mu.Lock()
	s.starts[path] = announcements{offset: advanced, found: found}
	s.mu.Unlock()
	return found
}

func contentBlocks(rec object) []object {
	message, ok := rec.object("message")
	if !ok {
		return nil
	}
	blocks, _ := message.objects("content")
	return blocks
}

func isASCIIAlnum(c byte) bool {
	return c >= '0' && c <= '9' || c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z'
}

// oneLine is a command as one line: its newlines become the spaces they read
// as, since it goes into a row one line tall.
func oneLine(text string) string {
	parts := strings.FieldsFunc(text, isNewline)
	kept := parts[:0]
	for _, p := range parts {
		if t := trimSpaces(p); t != "" {
			kept = append(kept, t)
		}
	}
	return strings.Join(kept, " ")
}

func isNewline(r rune) bool {
	switch r {
	case '\n', '\r', '\v', '\f', 0x85, 0x2028, 0x2029:
		return true
	}
	return false
}

// trimSpaces trims what Foundation calls whitespace without newlines: tabs
// and the space separators.
func trimSpaces(s string) string {
	return strings.TrimFunc(s, func(r rune) bool { return r == '\t' || unicode.In(r, unicode.Zs) })
}

// clipped keeps a line to what a phone-width row can show. `curl` of a web
// page prints a hundred kilobytes without a newline.
func clipped(line string) string {
	if utf8.RuneCountInString(line) <= shellRoom {
		return line
	}
	runes := []rune(line)
	return string(runes[:shellRoom-1]) + "…"
}

const (
	// MaxShellOutput is the most of one command's output a reader is sent:
	// its tail, never more. The Cloud word pins the same window
	// (`internal/app/cloudops`, the `shell` op).
	MaxShellOutput = 1 << 20
	// ShellOutputFloor is the least; a smaller ask is raised to it.
	ShellOutputFloor = 1 << 10
	// ShellOutputDefault is what a reader that names no window gets.
	ShellOutputDefault = 64 << 10
	// maxShellID is the longest id taken from a route. Claude Code's are nine
	// characters; this only keeps a hostile one from being a long string.
	maxShellID = 64
)

// ShellOutput is one background command's output as the Shell panel reads it
// — the Swift app's `Shells.output(of:id:bytes:)`, and its answer's four
// fields.
type ShellOutput struct {
	// Shell is the command's row: what started it and when it last printed.
	// Doing is empty once it has ended, as it is on the session row.
	Shell session.Shell
	// Text is the tail of the file, starting at a line boundary when it was
	// cut.
	Text string
	// Truncated is whether the file holds more than Text, before it.
	Truncated bool
	// Ended is whether the last line written is Claude Code's ending marker.
	Ended bool
	// Signature changes whenever the file does, so a reader repaints only when
	// bytes moved.
	Signature string
}

// ValidShellID is whether id can name an output file in the folder and
// nothing else: ASCII letters and digits only, which is every id Claude Code
// has minted on this machine (`b0aau3e6s`) and the only characters the
// announcement parser above accepts.
func ValidShellID(id string) bool {
	if id == "" || len(id) > maxShellID {
		return false
	}
	for i := 0; i < len(id); i++ {
		if !isASCIIAlnum(id[i]) {
			return false
		}
	}
	return true
}

// Output is the tail of one background command's output, at most window
// bytes of it.
//
// Only an id the transcript announced as a background command is read, the
// same rule Running keeps — but whether or not it has ended, because a reader
// watching a command must still see how it finished. Anything else, and an
// announced id whose file is gone, is not found.
func (s *Shells) Output(transcriptPath, id string, window int64) (ShellOutput, bool) {
	if !ValidShellID(id) {
		return ShellOutput{}, false
	}
	folder := s.Folder(transcriptPath)
	if folder == "" {
		return ShellOutput{}, false
	}
	if _, err := os.Stat(transcriptPath); err != nil {
		return ShellOutput{}, false
	}
	was, ok := s.announced(transcriptPath)[id]
	if !ok {
		return ShellOutput{}, false
	}
	file := filepath.Join(folder, id+".output")
	info, err := os.Stat(file)
	if err != nil || !info.Mode().IsRegular() {
		return ShellOutput{}, false
	}
	window = min(max(window, ShellOutputFloor), MaxShellOutput)
	data, whole, err := tailData(file, window)
	if err != nil {
		return ShellOutput{}, false
	}
	out := ShellOutput{
		Shell:     session.Shell{ID: id, At: info.ModTime(), Command: was.command, What: was.what},
		Text:      string(data),
		Truncated: !whole,
		Signature: fmt.Sprintf("%d-%d", info.ModTime().UnixNano(), info.Size()),
	}
	last := ""
	for _, line := range strings.Split(out.Text, "\n") {
		if t := trimSpaces(line); t != "" {
			last = t
		}
	}
	switch {
	case last == "":
	case isEnding(last):
		out.Ended = true
	default:
		out.Shell.Doing = clipped(last)
	}
	return out, true
}
